### ISSUE TITLE
404 errors should be harmonized

### ISSUE THREAD

**Issue by Hlavtox** [PrestaShop/PrestaShop#36857]
### What is wrong
Depending on type of the object, there are multiple ways 404 errors are handled for objects that don't exist.

### Redirect vs 404 template
- Manufacturer and Supplier controllers are redirecting to 404 controller, if the object can't be loaded.
- Category and Product controllers are staying on the current page, but setting a 404 template to be used.

### Different look of pages
For products, it's showing broken breadcrumbs and an error.
URL stays https://www.domain.cz/categoryrewrite/123-productrewrite.html

![product](https://github.com/user-attachments/assets/8b7a3fc5-ba78-4de1-a1fe-9a3b4bf358be)

For categories, only the message.
URL stays https://www.domain.cz/123-categoryrewrite

![category](https://github.com/user-attachments/assets/0e9395de-e81e-4e8f-bb9c-c9fa2a788ba4)

For 404 controller, only the message.
Page is redirected to URL https://www.domain.cz/index.php?controller=404

![404](https://github.com/user-attachments/assets/d11e002d-fd0e-4e70-a157-29acfeba9770)

### TODO
- Unify the way 404 errors are handled. Redirect or template changes?
- Setup a rewrite for 404 error controller.
- Verify that all ways to get to 404 errors behave the same.
  - A completely nonsense URL that can't be matched by the dispatcher.
  - Objects that don't exist.
  - Objects that are disabled or inaccessible.

**Comment by maniT4c** [comment]
Hello, 
for SEO, it's important to keep the url instead of redirect to the index.php?controller=404 page.
And I think the breadcrumb is useless, except if we are able to know in which category the page was, but I think it's not possible if the object doesn't exist anymore.

**Comment by pvujic** [comment]
> Hello, for SEO, it's important to keep the url instead of redirect to the index.php?controller=404 page. And I think the breadcrumb is useless, except if we are able to know in which category the page was, but I think it's not possible if the object doesn't exist anymore.

If a category was deleted tracing the category back will be impossible, but the breadcrumb would be useful in case the object is disabled or inaccessible.

In case of a product we could check the ID in the URL to get the category.

**Comment by Hlavtox** [comment]
@maniT4c @pvujic 

Yes, I agree. I did some research and apparently keeping the URL with a 404 response is a better solution than a redirect to `PageNotFoundController`, because it makes the errors harder to track and detect.

One downside is that if we set the 404 template like category and product controller does (`$this->setTemplate('errors/404')`), it still runs many of the code of the controller, hooks etc. And we have to settle for things like `$this->notfound` etc.

It would be best if we were able to run the `PageNotFoundController` controller directly, but keeping the URL.

> And I think the breadcrumb is useless

Yeah, that's just one of the bugs and incosistencies that I wanted to show.

> If a category was deleted tracing the category back will be impossible, but the breadcrumb would be useful in case the object is disabled or inaccessible.

We don't have to worry about tracing the category. If we can't find any content for the URL, just leave it be. :-)
